OpenRA.Mods.Common/AI/BaseBuilder.cs

```diff
@@ -37,14 +43,54 @@ public BaseBuilder(HackyAI ai, string category, Player p, PowerManager pm, Playe
 			playerPower = pm;
 			playerResources = pr;
 			this.category = category;
+			failRetryTicks = ai.Info.StructureProductionResumeDelay;
 		}
 
 		public void Tick()
 		{
+			// If failed to place something N consecutive times, wait M ticks until resuming building production
+			if (failCount >= ai.Info.MaximumFailedPlacementAttempts && --failRetryTicks <= 0)
+			{
+				var currentBuildings = world.ActorsWithTrait<Building>()
+					.Where(a => a.Actor.Owner == player)
+					.Count();
+
+				var baseProviders = world.ActorsWithTrait<BaseBuilding>()
+					.Where(a => a.Actor.Owner == player)
+					.Count();
+
+				// Only bother resetting failCount if either a) the number of buildings has decreased since last failure M ticks ago,
+				// or b) number of BaseProviders (construction yard or similar) has increased since then.
+				// Otherwise reset failRetryTicks instead to wait again.
+				if (currentBuildings < cachedBuildings || baseProviders > cachedBases)
+					failCount = 0;
+				else
+					failRetryTicks = ai.Info.StructureProductionResumeDelay;
+			}
+
 			// Only update once per second or so
 			if (--waitTicks > 0)
 				return;
 
+			if (!checkedWater)
+			{
+				waterAvailable = ai.EnoughWaterToBuildNaval();
```

Comment:
Managing the state of two flags feels kind of gross.  Can we use an enum with three values instead?